errors/errors.go

```diff
@@ -0,0 +1,220 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package errors
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+const (
+	ErrorCodesFeatureMetadataKey = "error_codes_feature"
+	Owner                        = "components-contrib"
+	Domain                       = "dapr.io"
+	unknown                      = "UNKNOWN_REASON"
+)
+
+const (
+	// gRPC to HTTP Mapping: 500 Internal Server Error
+	unknownHTTPCode = http.StatusInternalServerError
+)
+
+var UnknownErrorReason = WithErrorReason(unknown, unknownHTTPCode, codes.Unknown)
+
+type ResourceInfo struct {
+	Type string
+	Name string
+}
+
+// call this function to apply option
+type Option func(*Error)
+
+type Error struct {
+	err            error
+	description    string
+	reason         string
+	httpCode       int
+	grpcStatusCode codes.Code
+	metadata       map[string]string
+	resourceInfo   *ResourceInfo
+}
+
+// EnableComponentErrorCode stores an indicator for
+// components to determine if the ErrorCodes Feature
+// is enable.
+func EnableComponentErrorCode(metadata map[string]string) {
+	if metadata != nil {
+		metadata[ErrorCodesFeatureMetadataKey] = "true"
+	}
+}
+
+func featureEnabled(metadata map[string]string) bool {
+	_, ok := metadata[ErrorCodesFeatureMetadataKey]
+	return ok
+}
+
+// New create a new Error using the supplied metadata and ErrorOptions
+// **Note**: As this code is in `Feature Preview`, it will only continue processing
+// if the ErrorCodes is enabled
+// TODO: @robertojrojas update when feature is ready.
+func New(err error, metadata map[string]string, options ...Option) *Error {
+	// The following condition can be removed once the
+	// Error Codes Feature is GA
+	if !featureEnabled(metadata) {
+		return nil
+	}
+
+	if err == nil {
+		return nil
+	}
+
+	// Use default values
+	de := &Error{
+		err:            err,
+		reason:         unknown,
+		httpCode:       unknownHTTPCode,
+		grpcStatusCode: codes.Unknown,
+	}
+
+	// Now apply any requested options
+	// to override
+	for _, option := range options {
+		option(de)
+	}
+
+	return de
+}
+
+// Error implements the error interface.
+func (e *Error) Error() string {
+	if e != nil && e.err != nil {
+		return e.err.Error()
+	}
+	return ""
+}
+
+// Unwrap implements the error unwrapping interface.
+func (e *Error) Unwrap() error {
+	return e.err
+}
+
+// Description returns the description of the error.
+func (e *Error) Description() string {
+	if e.description != "" {
+		return e.description
+	}
+	return e.err.Error()
+}
+
+func WithErrorReason(reason string, httpCode int, grpcStatusCode codes.Code) Option {
+	f := func(er *Error) {
+		er.reason = reason
+		er.grpcStatusCode = grpcStatusCode
+		er.httpCode = httpCode
+	}
+	return f
```

Comment:
nit: no need to assign a variable which gets returned, and we should call the error variable `err` as this is consistent in Go. Same applies to for the other funcs below.

```suggestion
	return func(err *Error) {
		err.reason = reason
		err.grpcStatusCode = grpcStatusCode
		err.httpCode = httpCode
	}
```